### ISSUE TITLE
Modules draft, version 2

### ISSUE THREAD

**Issue by christianalfoni** [cerebral/cerebral#158]
Okay, so I have been thinking :-) The concept for modules has been thought of as an addition to the current API of Cerebral, but I think we have to rethink a bit more to get this nicely fit in with Cerebral.

### Remove top level state, signals and services
First of all I suggest we remove defining state as part of the Model. `controller.signal` is neither normally used, it is "lower level". I also want to remove passing in services to the controller. That will give us:

```js
import Controller from 'cerebral';
import Model from 'cerebral-baobab';

const model = Model({}); // can add top level state if you want to
const controller = Controller(model);
```

But no services are defined here.

### Everything is a module
So to even get going with Cerebral you have to define a module, but instead of passing the controller, we pass an actual module where you set state, register signals and services.

```js
import Controller from 'cerebral';
import Model from 'cerebral-baobab';
import Home from './modules/Home';

const model = Model({});
const controller = Controller(model);

controller.registerModules({
  home: Home({foo: 'bar'})
});
```
And the module is defined with a function receiving the module. On the module you have access to the name, its options, signals, registering signals and services.

```js
export default (options = {}) => {
  return (module, controller) => {
    module.name // "home"
    module.signals // {}
    
    // Add alias to module for sharing, look below
    module.alias('christianalfoni/mymodule');
  
    // Add initial namespaced state
    module.state({
      title: 'Home'
    });

    // Register a namespaced signal
    module.signal('buttonClicked', chain);
    module.signalSync('inputChanged', chain);

    // Register a namespaced service
    module.service('interval', () => {});
  
    // Register submodules
    module.registerModules({
      mySubmodule: Submodule()
    });

    // Return any meta data
    return {};
  }
}
```

Controller is exposed for granular control and ability to trigger signals on other modules.

### Router becomes a module

```js
import Controller from 'cerebral';
import Model from 'cerebral-baobab';
import Router from 'cerebral-router';
import Home from './modules/Home';

const model = Model({});
const controller = Controller(model);

const modules = controller.modules({
  home: Home({foo: 'bar'}),
  router: Router({
    '/': 'home.mounted'
  })
});
```

### Signals are only registered on controller
```js
// React
render() {
  const {signals} = this.props;
  return <button onClick={() => signals.home.buttonClicked()}>Click</button>
}
```

### Sharing modules
When you create your module for sharing you give it a unique name, something the depending components/actions etc. can use to find your module.

```js
export default (options = {}) => {
  return (module, controller) => {
    module.alias('christianalfoni/mymodule');
    return {};
  }
}
```

That means this alias just points to the module created with a custom name. So:

```js
@Cerebral((props) => {
  const modulePath = props.modules['christianalfoni/mymodule'].path;
  return {
    moduleState: [modulePath, 'someState']
  }
})
class Comp extends React.Component {
  render() {
    const moduleName = this.props.modules['christianalfoni/mymodule'].name;
    const {signals} = this.props.signals[moduleName];
    return <button onClick={() => signals.buttonClicked()}>Click</button>
  }
}
```

This means that modules are also exposed to actions, but they have some extra helpers:

```js
function myAction({module}) {
  module.state.set([moduleName, 'sometState'], 'bar']);
  module.services[moduleName].doSomething();
}
```

This operates on the services and state of the module the action is triggered on. You can also point to other modules:

 ```js
function myAction({modules}) {
  modules['christianalfoni/mymodule'].state.set([moduleName, 'sometState'], 'bar']);
  modules.app.state.set(['foo'], 'bar');
}
```
What about multiple instance of modules? Still same. You pass in the name of the module the component/action should work on, as it needs to know the specific instance anyways.

### Types of modules
Just some pseudo code to show how we can use it.

#### cerebral-ajax
```js
import superagent from 'superagent';

export default (options = {}) => {
  return (module) => {
    module.service('get', () => return new Promise(...));
  };
}

// Used as
function myAction({services}) {
  services.ajax.get()
}
```

#### cerebral-ui
```js
export default (options = {}) => {
  return (module) => {
    module.alias('cerebral-module-ui');
    module.signalSync('inputChanged', chain);
  }
}

// Used as
<Input sync={['someModule', 'somePath']}/>
```

#### cerebral-firebase
```js
import Firebase from 'firebase';

export default (options = {}) => {
  return (module, controller) => {
    const alias = 'cerebral/firebase';
    module.alias(alias);
    module.service('sync', (path) => {
      const ref = new Firebase(options.url + path);
      ref.on('value', () => module.signals.valueUpdated)
    });
    module.signal('valueUpdated', chain({
      path: options.path
    }));
  };
}

// Used as
function myAction({services}) {
  services.firebase.sync('somepath');
}
```

### Access module namespace on actions
As suggested by... hm... can not find it, anyways :-) Expose module name on actions to more easily point to correct path.

```js
function myAction({state, module}) {
  state.get([module, 'todos']);
}
```

This ensures that you can grab any state, but more easily get the state related to which module the signal is registered to.

### Summary
So the goal here is to force use of modules, because they are great! But we have to clean up the API and make it more approachable, which I think this suggestion does.

**Comment by Guria** [comment]
Would bethere ability to define 'global' not scopes one?

**Comment by bfitch** [comment]
> Everything is a Module

Yes! I think this really crystalizes the main idea behind modules. A cerebral app is composed of one to many modules, just like UI is composed of one to many components. Actually...modules are kinda like a cerebral extension to components...they could even be called Cerebral Components :smile:

**Comment by garth** [comment]
Do we still pass modules to component render?

**Comment by christianalfoni** [comment]
@garth Jup, sorry, I will add that :-)

**Comment by christianalfoni** [comment]
Okay, updated

**Comment by christianalfoni** [comment]
Okay, I will make a branch of this. I got a bit excited :-D

**Comment by garth** [comment]
Looks good to me. Would this be for cerebral 2.0 or will you hold off on 1.0 until this stuff is ready?

**Comment by christianalfoni** [comment]
Looking into it now, does not have to be breaking change. Will create a video when done, we can discuss it :-)

**Comment by garth** [comment]
Only a breaking change for apps that have started to use the new module syntax (which was experimental anyway)

**Comment by delaneyj** [comment]
>Remove top level state, signals and services

@garth wouldn't that be a breaking change for everyone?

**Comment by abalmos** [comment]
Much cleaner! Writing modules is now basically just writing a mini cerebral app. :+1: 

Currently I have my own simple module like thing where a "module" exposes react components and a `register(controller)` method. The main controller.js file calls all the `register` functions with the current controller and then the "module" can add anything it wants to the cerebral instance. I haven't really had a need for more than that up to this point. That said, I much prefer your suggestion of `controller.register(module)` and I very much appreciate the auto-prefix of the module name.

**Alias**
 
I am having a hard time seeing the advantage of `alias()` over just global names. The alias still needs to be unique to locate the module and if the alias is unique why not just register the module with the alias name and be done with it? The module function doesn't normally need to know its own name because `module.signal()`, `module.signalSync()`, and `module.service()` will auto-prefix it. If for some reason a module does need it's own name it can be had from `module.name`.

**`module.name`**
A time you might need `module.name` is when a module calls it's own signals. From @christianalfoni's Firebase example (alias replaced with module.name):

```js
    module.service('sync', (path) => {
      const ref = new Firebase(options.url + path);
      ref.on('value', () => controller.signals[module.name].valueUpdated)
    });
```
We could expose `modules.signals`, where `modules.signals = controller.signals[module.name]`. 
Then we would have:

```js
    module.service('sync', (path) => {
      const ref = new Firebase(options.url + path);
      ref.on('value', () => module.signals.valueUpdated)
    });
```

I suppose the same for `module.services`. 

That would complete the "modules are mini cerebral apps" because we would have brought the entire cerebral api (I think) into the "auto-prefixed" `modules`.

**Final thoughts on naming**
I am probably missing something important but I think we might be worrying about names too much. As @christianalfoni pointed out Angluar, and even npm, etc, have let the module writer define the module's global name and that seems to be working out pretty well. Sometimes you get stuff like: `koa-body`, `koa-bodyparser`, `koa-body-parser` but it's not that often and typically the community eventually settles on one of them.

Right now my vote would go for everything @christianalfoni proposes expect `alias()` (I say lets wait until we have naming issues) and `modules.signals`/`modules.services`.

**Side thought: npm names for modules**
I propose borrow from Babel and encourage module writers to name their cerebral module like: cerebral-module-{name} when publishing to npm. Not only will the help avoid module name clashes it also makes finding cerebral modules easier. 

https://www.npmjs.com/search?q=cerebral-module-

**Comment by Guria** [comment]
@delaneyj of course it would. But until 1.0.0 every 0.x release could have breaking changes.

**Comment by delaneyj** [comment]
@Guria  good point.  All of this would really clean up cerebral even more than it around is compared to other options.  Being very new to cerebral @abalmos' suggestion makes perfect sense.  I wonder if things like snabbdom-material would also live in a module or is that going to far?  I mean they are 'just' function.

**Comment by garth** [comment]
@delaneyj @Guria I think that it would still be possible to use cerebral without modules, though the non-module approach may be deprecated after a period of stabilisation. So the breaking changes would only affect existing modules, which since modules have only been around for a couple of weeks, wouldn't be too many people/projects.

**Comment by garth** [comment]
@abalmos module naming has been the biggest problem to solve. Within your own app it would almost never be an issue. it's only a concern when as a shared module implementer you don't know what apps will call your module. the alias would only be for internal use within the module. A name that is extremely unlikely to clash with other modules, and should never need to be publicly exposed.

As a module publisher my module may have 4 or 5 different parts. The module itself, Components, Actions, Signals and Services. I need to be sure that my service is talking to the correct module, my UI components are talking to the correct module etc... If I don't control the naming of the module I can't guarantee that everything will work seamlessly together.

Most modules will never need to use an alias, but it's important that it's there.

**Comment by garth** [comment]
@delaneyj component libraries probably don't need to be loaded as modules and are well suited to regular imports/requires. Though there may be cases where they might contain a companion module.

A good example might be where you have a `/screen` element in your state tree that knows about the physical characteristics of the device (portrait/landscape, phone/tablet/desktop, etc...). A module would be an ideal candidate to manage this state and ensure that it's always current. Components in the library can then read this state and adjust their presentation/behaviour accordingly.

**Comment by abalmos** [comment]
@garth Re-reading my comment, it is clear that I did not achieve my goal of presenting two ways of thinking. Let me try to say what I really wanted to: :smile:

Two ways of thinking:

1. Keep `alias()` and **allow** applications to rename modules at registration time. There's nothing wrong with the approach, I just see it as complicated. However, inside of the module definition things could be made a little more consistent with the controller by exposing a 'modules.signals' and 'modules.services' that are namespaced to whatever the runtime name of the module is.

2. Drop `alias()` and **not allow** applications to rename modules. If modules had fixed and global names picked by their writer then the naming issues go away. Seems like the whole issue can be solved by encouraging a decent naming scheme:
  - Local modules that are not shared with the world are named something very unlikely to collide with a public module.
  - Public modules will typically be distributed via npm and so the name uniqueness requirement of npm will lead to mostly unique module names. Node modules have the same name problems but so far it is not been an overwhelming issue for the community.

If it's not clear I would vote for 2 but happy with 1. I would really like to see this make v1.0 so I don't want to push too many buttons here.

**Comment by tnrich** [comment]
@abalmos I remember hearing something about needing to provide different local names for "multiple" instances of the same module to avoid clashes. Maybe that is why @christianalfoni is suggesting the `alias()` approach? I'm not sure though...

**Comment by abalmos** [comment]
@tnrich Hmmm... Good point. That's likely to come up with back end services.

**Comment by delaneyj** [comment]
> @garth  component libraries probably don't need to be loaded as modules and are well suited to regular imports/requires. 

Which why I was wondering if that's going too far.  On the flip side, once you start having things like <Link/> in a component library that would tie to something like signals.pageChange then it starts to make more sense.

**Comment by Guria** [comment]
@abalmos you can assign any lib required from `npm` to any variable name since usually it completely scoped. So should be with most cerebral-modules. You can name it while publishing, but app author could name it by own needs in his app.

**Comment by abalmos** [comment]
@Guria That's true but I think here module name corresponds with the package name. It seems the main issue is when you need to refer to the same module across two or more files. That's when you have to rely on knowing the package name.

That said, what @tnrich mentioned is probably justification for `alias()`. We should make sure to document the reasons one would want to use `alias()`. 

I appreciate everyone's help!

**Comment by garth** [comment]
A pouchdb module is a good example of one that requires multiple instances.

**Comment by tnrich** [comment]
@abalmos I'd want to make sure that what I said above is actually a real issue and reason for having alias()

**Comment by christianalfoni** [comment]
Hi guys :-) Great discussion!

It would be great to use global names, but they will just be too much work using. Like:

```js
this.props.signals['cerebral-module-whatever'].buttonClicked();
```

You want short names for modules to work with them, but still ensure that modules can talk to each other regardless of this shortname :-)

About exposing signals on the `module` when creating it I agree on, makes it more clean :-) That said, I do not think we should expose signals/services on the modules property passed to actions and components.

```js
function myAction({modules}) {
  modules.myModule.services.something();
}

render() {
  <button onClick={() => this.props.modules.myModule.clicked()}>Click</button>
}
```

The reasons are:

1. It would be the same module object passed into actions and components. But you can not call signals from actions and you can not call services from components. It would be really weird and technically complex to send in one version of the module to actions and a different version to components

2. It is shorter to write `this.props.signals.myModule.clicked()` than `this.props.modules.myModule.signals.clicked()`

3. We should only have one way to do it, to avoid confusion and the modules version does not work for current applications, but using signals/services works for current applications and modules based applications

**Comment by christianalfoni** [comment]
Okay, uploaded a video: https://www.youtube.com/watch?v=yiaZmhLC8MA

**Comment by NervosaX** [comment]
Does this system allow for nested modules? Currently one of my modules would be so big, that .signals would return 20-30 methods to call. It would be great to be able to register modules inside other modules:

`this.props.signals.myModules.subModule.clicked()`

This could be great with splitting up routing and state as well across apps. Say that I have a complex system to create/edit/list/detail an email (very naively).

I could call the module "email". The 4 views listed are so complex that combining all of its functionality into the email module would be (hypothetically) very hard to reason. Perhaps it has 30 signals, lots of actions specific to each view. Perhaps some common functionality, but mostly separate to each type of view. It would be better to split this up into sub modules:

```
- email
  - create
  - edit
  - list
  - detail
```

Perhaps email is a module. It's route is "/email";
create, edit and detail have a route like ":id/create". I have a use case that would require 3-4 routes for one or two of these types of views, so I don't feel that that is too silly an idea.

Each has their own state, which will be based off the previous state:

```
{
  email: {
    create: {},
    edit: {},
    list: {},
    detail: {}
  }
} 
```

and signals can be specific to their submodule:
`signals.email.create.formSubmitted()`
`signals.email.list.pageOpened()`

I can currently create a signal with a format like the above, so it feels like modules should be able to cope with this kind of complex structure too?

**Comment by christianalfoni** [comment]
Investigating the following syntax now:

```js
export default (options = {}) {
  return (module) => {
    
    module.state({foo: 'bar'});

    module.registerModules({
      submodule: SomeModule()
    });

  }
}
```

**Comment by NervosaX** [comment]
Can this be closed?

**Comment by dwiyatci** [comment]
> I suggest we remove defining state as part of the Model.

@christianalfoni Is this basically saying you revert your suggestion of having one giant (Baobab) state tree of our application in only one file? (I'm pretty sure I heard it quite often when I was watching some of early Cerebral videos in your YouTube channel :s)

**Comment by christianalfoni** [comment]
@dwiyatci Almost :-) If you have a rather small application you can still put everything inside one single module. But now you have the ability to split things up if you need to. But the huge thing here is the possibility to share modules on npm and empower other Cerebral developers :-)

**Comment by dwiyatci** [comment]
:+1:
